### PR TITLE
[python] Experiment-level upgrader/resizer

### PR DESCRIPTION
**Pull request by johnkerl**
**Issue and/or context:** As tracked on issue #2407 / [[sc-51048]](https://app.shortcut.com/tiledb-inc/story/51048).

Note that the intended Python and R API changes are all agreed on and finalized as described in #2407.

**Changes:**

Tests the experiment-level upgrader and resizer.

**Notes for Reviewer:**

There is more to do in `tiledbsoma.io` for append-mode ingest on #3148.

Example verbose output:

```
uri = "/var/ns/pbmc3k"
tiledbsoma.io.show_experiment_shapes(uri)
```

<details>

```
[DataFrame] obs
  URI file:///var/ns/pbmc3k/obs
  count                2638
  domain               ((0, 9223372036854773758),)
  maxdomain            ((0, 9223372036854773758),)
  upgraded             False

[DataFrame] ms/RNA/var
  URI file:///var/ns/pbmc3k/ms/RNA/var
  count                1838
  domain               ((0, 9223372036854773758),)
  maxdomain            ((0, 9223372036854773758),)
  upgraded             False

[SparseNDArray] ms/RNA/X/data
  URI file:///var/ns/pbmc3k/ms/RNA/X/data
  used_shape           ((0, 2637), (0, 1837))
  shape                (9223372036854773759, 9223372036854773759)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             False

[SparseNDArray] ms/RNA/obsm/X_draw_graph_fr
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_draw_graph_fr
  used_shape           ((0, 2637), (0, 1))
  shape                (9223372036854773759, 9223372036854773759)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             False

[SparseNDArray] ms/RNA/obsm/X_pca
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_pca
  used_shape           ((0, 2637), (0, 49))
  shape                (9223372036854773759, 9223372036854773759)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             False

[SparseNDArray] ms/RNA/obsm/X_tsne
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_tsne
  used_shape           ((0, 2637), (0, 1))
  shape                (9223372036854773759, 9223372036854773759)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             False

[SparseNDArray] ms/RNA/obsm/X_umap
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_umap
  used_shape           ((0, 2637), (0, 1))
  shape                (9223372036854773759, 9223372036854773759)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             False

[SparseNDArray] ms/RNA/obsp/connectivities
  URI file:///var/ns/pbmc3k/ms/RNA/obsp/connectivities
  used_shape           ((0, 2637), (0, 2637))
  shape                (9223372036854773759, 9223372036854773759)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             False

[SparseNDArray] ms/RNA/obsp/distances
  URI file:///var/ns/pbmc3k/ms/RNA/obsp/distances
  used_shape           ((0, 2637), (0, 2637))
  shape                (9223372036854773759, 9223372036854773759)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             False

[SparseNDArray] ms/RNA/varm/PCs
  URI file:///var/ns/pbmc3k/ms/RNA/varm/PCs
  used_shape           ((0, 1837), (0, 49))
  shape                (9223372036854773759, 9223372036854773759)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             False

[DataFrame] ms/raw/var
  URI file:///var/ns/pbmc3k/ms/raw/var
  count                13714
  domain               ((0, 9223372036854773758),)
  maxdomain            ((0, 9223372036854773758),)
  upgraded             False

[SparseNDArray] ms/raw/X/data
  URI file:///var/ns/pbmc3k/ms/raw/X/data
  used_shape           ((0, 2637), (0, 13713))
  shape                (9223372036854773759, 9223372036854773759)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             False
```

</details>

```
uri = "/var/ns/pbmc3k"
tiledbsoma.io.upgrade_experiment_shapes(uri, verbose=True)
```

<details>

```
[DataFrame] obs
  URI file:///var/ns/pbmc3k/obs
  Applying upgrade_soma_joinid_shape(2638)

[DataFrame] ms/RNA/var
  URI file:///var/ns/pbmc3k/ms/RNA/var
  Applying upgrade_soma_joinid_shape(1838)

[SparseNDArray] ms/RNA/X/data
  URI file:///var/ns/pbmc3k/ms/RNA/X/data
  Applying upgrade_shape((2638, 1838))

[SparseNDArray] ms/RNA/obsm/X_draw_graph_fr
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_draw_graph_fr
  Applying upgrade_shape((2638, 2))

[SparseNDArray] ms/RNA/obsm/X_pca
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_pca
  Applying upgrade_shape((2638, 50))

[SparseNDArray] ms/RNA/obsm/X_tsne
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_tsne
  Applying upgrade_shape((2638, 2))

[SparseNDArray] ms/RNA/obsm/X_umap
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_umap
  Applying upgrade_shape((2638, 2))

[SparseNDArray] ms/RNA/obsp/connectivities
  URI file:///var/ns/pbmc3k/ms/RNA/obsp/connectivities
  Applying upgrade_shape((2638, 2638))

[SparseNDArray] ms/RNA/obsp/distances
  URI file:///var/ns/pbmc3k/ms/RNA/obsp/distances
  Applying upgrade_shape((2638, 2638))

[SparseNDArray] ms/RNA/varm/PCs
  URI file:///var/ns/pbmc3k/ms/RNA/varm/PCs
  Applying upgrade_shape((1838, 50))

[DataFrame] ms/raw/var
  URI file:///var/ns/pbmc3k/ms/raw/var
  Applying upgrade_soma_joinid_shape(13714)

[SparseNDArray] ms/raw/X/data
  URI file:///var/ns/pbmc3k/ms/raw/X/data
  Applying upgrade_shape((2638, 13714))
```

</details>

```
uri = "/var/ns/pbmc3k"
tiledbsoma.io.resize_experiment(uri, 2638, {"RNA": 5000, "raw": 13714})
```

<details>

```
[DataFrame] obs
  URI file:///var/ns/pbmc3k/obs
  Applying resize_soma_joinid_shape(2638)

[DataFrame] ms/RNA/var
  URI file:///var/ns/pbmc3k/ms/RNA/var
  Applying resize_soma_joinid_shape(5000)

[SparseNDArray] ms/RNA/X/data
  URI file:///var/ns/pbmc3k/ms/RNA/X/data
  Applying resize((2638, 5000))

[SparseNDArray] ms/RNA/obsm/X_draw_graph_fr
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_draw_graph_fr
  Applying resize((2638, 2))

[SparseNDArray] ms/RNA/obsm/X_pca
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_pca
  Applying resize((2638, 50))

[SparseNDArray] ms/RNA/obsm/X_tsne
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_tsne
  Applying resize((2638, 2))

[SparseNDArray] ms/RNA/obsm/X_umap
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_umap
  Applying resize((2638, 2))

[SparseNDArray] ms/RNA/obsp/connectivities
  URI file:///var/ns/pbmc3k/ms/RNA/obsp/connectivities
  Applying resize((2638, 2638))

[SparseNDArray] ms/RNA/obsp/distances
  URI file:///var/ns/pbmc3k/ms/RNA/obsp/distances
  Applying resize((2638, 2638))

[SparseNDArray] ms/RNA/varm/PCs
  URI file:///var/ns/pbmc3k/ms/RNA/varm/PCs
  Applying resize((5000, 50))

[DataFrame] ms/raw/var
  URI file:///var/ns/pbmc3k/ms/raw/var
  Applying resize_soma_joinid_shape(13714)

[SparseNDArray] ms/raw/X/data
  URI file:///var/ns/pbmc3k/ms/raw/X/data
  Applying resize((2638, 13714))
```

</details>

```
uri = "/var/ns/pbmc3k"
tiledbsoma.io.show_experiment_shapes(uri)
```

<details>

```
[DataFrame] obs
  URI file:///var/ns/pbmc3k/obs
  count                2638
  domain               ((0, 2637),)
  maxdomain            ((0, 9223372036854773758),)
  upgraded             True

[DataFrame] ms/RNA/var
  URI file:///var/ns/pbmc3k/ms/RNA/var
  count                1838
  domain               ((0, 4999),)
  maxdomain            ((0, 9223372036854773758),)
  upgraded             True

[SparseNDArray] ms/RNA/X/data
  URI file:///var/ns/pbmc3k/ms/RNA/X/data
  used_shape           ((0, 2637), (0, 1837))
  shape                (2638, 5000)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             True

[SparseNDArray] ms/RNA/obsm/X_draw_graph_fr
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_draw_graph_fr
  used_shape           ((0, 2637), (0, 1))
  shape                (2638, 2)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             True

[SparseNDArray] ms/RNA/obsm/X_pca
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_pca
  used_shape           ((0, 2637), (0, 49))
  shape                (2638, 50)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             True

[SparseNDArray] ms/RNA/obsm/X_tsne
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_tsne
  used_shape           ((0, 2637), (0, 1))
  shape                (2638, 2)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             True

[SparseNDArray] ms/RNA/obsm/X_umap
  URI file:///var/ns/pbmc3k/ms/RNA/obsm/X_umap
  used_shape           ((0, 2637), (0, 1))
  shape                (2638, 2)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             True

[SparseNDArray] ms/RNA/obsp/connectivities
  URI file:///var/ns/pbmc3k/ms/RNA/obsp/connectivities
  used_shape           ((0, 2637), (0, 2637))
  shape                (2638, 2638)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             True

[SparseNDArray] ms/RNA/obsp/distances
  URI file:///var/ns/pbmc3k/ms/RNA/obsp/distances
  used_shape           ((0, 2637), (0, 2637))
  shape                (2638, 2638)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             True

[SparseNDArray] ms/RNA/varm/PCs
  URI file:///var/ns/pbmc3k/ms/RNA/varm/PCs
  used_shape           ((0, 1837), (0, 49))
  shape                (5000, 50)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             True

[DataFrame] ms/raw/var
  URI file:///var/ns/pbmc3k/ms/raw/var
  count                13714
  domain               ((0, 13713),)
  maxdomain            ((0, 9223372036854773758),)
  upgraded             True

[SparseNDArray] ms/raw/X/data
  URI file:///var/ns/pbmc3k/ms/raw/X/data
  used_shape           ((0, 2637), (0, 13713))
  shape                (2638, 13714)
  maxshape             (9223372036854773759, 9223372036854773759)
  upgraded             True
```

</details>

### CODE REVIEW

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 1]
```diff
@@ -0,0 +1,546 @@
...
+    """
+
+    nobs: Optional[int]
+    nvars: Optional[Dict[str, int]]
+    ms_name: Optional[str]
+    coll_name: Optional[str]
+    verbose: bool
+    check_only: bool
+    context: Optional[tiledbsoma.SOMATileDBContext]
+    output_handle: Printable
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/shaping.py:33]: Would suggest importing the `attrs` module for using data classes. You can leverage default values with this.

```suggestion
@attr.s(auto_attribs=True, frozen=True)
class SizingArgs:
    nobs: Optional[int] = None
    nvars: Optional[Dict[str, int]] = None
    ms_name: Optional[str] = None
    coll_name: Optional[str] = None
    verbose: bool = True
    check_only: bool = True
    context: Optional[tiledbsoma.SOMATileDBContext] = None
    output_handle: Printable
```

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 2]
```diff
@@ -0,0 +1,546 @@
...
+        context=context,
+        output_handle=output_handle,
+    )
+
+    # Do an early check on the nvars keys vs the experiment's
+    # measurent names. This isn't a can-do status for the experiment;
+    # it's a failure of the user's arguments.
+    with tiledbsoma.Experiment.open(uri) as exp:
+        arg_keys = sorted(list(set(nvars.keys())))
+        ms_keys = sorted(list(set(exp.ms.keys())))
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/shaping.py:169]: Dictionary keys are already be unique, so I don't think the call to `set` is necessary.

```suggestion
        arg_keys = sorted(nvars.keys())
        ms_keys = sorted(exp.ms.keys())
```

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 3]
```diff
@@ -0,0 +1,546 @@
...
+                    args = dict(
+                        nobs=args["nobs"],
+                        nvars=args["nvars"],
+                        ms_name=args["ms_name"],
+                        coll_name=coll_name,
+                        verbose=args["verbose"],
+                        check_only=args["check_only"],
+                        context=args["context"],
+                        output_handle=args["output_handle"],
+                    )
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/shaping.py:224]: (I think you can do this)

```suggestion
                    args = dict(coll_name=coll_name, **args)
```

Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/shaping.py:224]: Or this

```suggestion
                    args[coll_name] = coll_name
```

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 4]
```diff
@@ -0,0 +1,546 @@
...
+                    args = dict(
+                        nobs=args["nobs"],
+                        nvars=args["nvars"],
+                        ms_name=key,
+                        coll_name=args["coll_name"],
+                        verbose=args["verbose"],
+                        check_only=args["check_only"],
+                        context=args["context"],
+                        output_handle=args["output_handle"],
+                    )
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/shaping.py:248]:
```suggestion
                    args[ms_name] = key
```

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 5]
```diff
@@ -0,0 +1,546 @@
...
+) -> str:
+    # Return things like "ms/RNA/X/data"
+    pieces = []
+    if args["ms_name"] is not None:
+        pieces.append("ms")
+        pieces.append(args["ms_name"])
+    if args["coll_name"] is not None:
+        pieces.append(args["coll_name"])
+    pieces.append(node_name)
+    return "[" + type_name + "] " + "/".join(pieces)
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/shaping.py:461]:
```suggestion
    return f"[{type_name}] {'/'.join(pieces)} "
```

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 6]
```diff
@@ -0,0 +1,546 @@
...
+
+
+def _bannerize(
+    args: SizingArgs,
+    name: str,
+    value: Any,
+) -> None:
+    if args["verbose"] or args["check_only"]:
+        padded_name = "%-20s" % name
+        print(f"  {padded_name} {value}", file=args["output_handle"])
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/shaping.py:471]:
```suggestion
        print(f"  {name:<20} {value}", file=args["output_handle"])
```

File: apis/python/src/tiledbsoma/io/shaping.py [hunk 7]
```diff
@@ -0,0 +1,546 @@
...
+    elif coll_name == "obsp":
+        return (nobs, nobs)
+    elif coll_name == "varm":
+        return (nvar, current_shape[1])
+    elif coll_name == "varp":
+        return (nvar, nvar)
+    else:
+        raise tiledbsoma.SOMAError(
+            f"experiment resize: internal error: unhandled collection {coll_name}"
+        )
```
Comment (nguyenv) [inline apis/python/src/tiledbsoma/io/shaping.py:546]:
```suggestion
    coll_dict = {
        "X": (nobs, nvar),
        "obsm": (nobs, current_shape[1]),
        "obsp": (nobs, nobs),
        "varm": (nvar, current_shape[1]),
        "varp": (nvar, nvar),
    }

    try:
        return coll_dict[coll_name]
    except KeyError:
        raise tiledbsoma.SOMAError(
            f"experiment resize: internal error: unhandled collection {coll_name}"
        )

```

**Comment by johnkerl** [comment]
@nguyenv I decided not to go with `attrs` over `TypedDict` -- sort of a po-tay-to po-tah-to ...

**Review by nguyenv — APPROVED** [review]
lgtm; Internally, there isn't much difference between using `attrs` vs `TypedDict`, so it is fine keeping it as is. I do think `attrs` does give us additional advantages like providing more enhanced type checking and default value generators which could simplify the code here but not substantial that it warrants changing.

**Pull request merged**